### PR TITLE
added spring security

### PR DESCRIPTION
**Pull request by vladshender**

### CODE REVIEW

File: src/main/java/com/example/controller/BookController.java [hunk 1]
```diff
@@ -41,19 +42,22 @@ public BookDto getBookById(@PathVariable Long id) {
         return bookService.getBookById(id);
     }
 
+    @PreAuthorize("hasAuthority('ADMIN')")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/BookController.java:45]: Завжди додавай ROLE_ при використанні методу hasAuthority(), наприклад, hasAuthority("ROLE_CUSTOMER").

See the lecture material

File: src/main/java/com/example/model/User.java [hunk 2]
```diff
@@ -32,4 +42,52 @@ public class User {
     private String lastName;
 
     private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/User.java:46]:
```suggestion
    @ManyToMany
```

File: src/main/java/com/example/model/User.java [hunk 3]
```diff
@@ -32,4 +42,52 @@ public class User {
...
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles = new HashSet<>();
+
+    @Column(name = "is_deleted", nullable = false)
+    private boolean isDeleted;
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return roles.stream()
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/User.java:59]:
```suggestion
        return roles;
```

File: src/main/java/com/example/security/CustomUserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(
+                        () -> new EntityNotFoundException("Can`t find user bu email: " + email)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/security/CustomUserDetailsService.java:21]:
```suggestion
                        () -> new EntityNotFoundException("Can`t find user by email: " + email)
```

File: src/main/java/com/example/service/user/UserServiceImpl.java [hunk 5]
```diff
@@ -24,6 +26,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
 
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(user.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/user/UserServiceImpl.java:29]: Should we set User role?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! See comments

**Comment by vladshender** [comment]
without FetchType.Eager, a LazyInitializationException is thrown

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**